feat(instrumentation-react-native-navigation) Adding new instrumentation library to track navigation for React Native apps (#13)
embrace-io/opentelemetry-js-contrib: plugins/node/instrumentation-react-native-navigation/README.md

# OpenTelemetry Navigation Instrumentation for React Native

[![NPM Published Version][npm-img]][npm-url]
[![Apache License][license-image]][license-image]

This module provides instrumentation for [react-native/nagivation](https://reactnavigation.org/docs/getting-started), [expo-router](https://docs.expo.dev/router/introduction/) and [wix/react-native-navigation](https://wix.github.io/react-native-navigation/docs/before-you-start/)

## Installation
```
npm i @opentelemetry/instrumentation-react-native-navigation @opentelemetry/api
```

or if you use yarn

```
yarn add @opentelemetry/instrumentation-react-native-navigation @opentelemetry/api
```

## Supported Versions
  - Nodejs `>=14`

## Usage

> This package is designed to streamline your workflow by requiring minimal setup. To use this package, you only need to pass a reference and a provider.

Reviewer: ```suggestion
This package is designed to streamline your workflow by requiring minimal setup. To use this package, you only need to pass a reference and a optionally provider (the global one will be used by default).
```

Author: fixed

If you are using `expo-router` or `react-native/navigation` you need to wrap your entire application with the `NavigationTracker` component.

```javascript
import {FC} from 'react';
import {Stack, useNavigationContainerRef} from 'expo-router';
import {NavigationTracker} from '@opentelemetry/instrumentation-react-native-navigation';

const App: FC = () => {
  const navigationRef = useNavigationContainerRef(); // if you do not use `expo-router` the same hook is also available in `@react-navigation/native` since `expo-router` is built on top of it. Just make sure this ref is passed also to the navigation container at the root of your app (if not, the ref would be empty and you will get a console.warn message instead).

  const provider = useProvider(); // the provider is something you need to configure and pass down as prop into the `NavigationTracker` component (this hook is not part of the package, it is just used here as a reference)

>   // If your choise is not to pass any custom tracer provider, the <NavigationTracker /> component will use the global one.

Reviewer: ```suggestion
  // If your choice is not to pass any custom tracer provider, the <NavigationTracker /> component will use the global one.
```

Author: fixed

  return (
    <NativeNavigationTracker ref={navigationRef} provider={provider}>
      {/* content of the app goes here */}
    </NavigationTracker>
  );
};

export default App;
```

## Goal
The purpose of this package is to intercept changes in the navigation of a React Native application and create telemetry data following the OpenTelemetry standards. Every new view displayed will start a new Span, which will end only when the next view becomes available to the user.

For instance, when the application starts and the user navigates to a new section, the first Span will be considered finished at that moment. Let’s take a look at the output of this Span:

```
{
  resource: {
    attributes: {
      'service.name': 'navigation',
      'telemetry.sdk.language': 'nodejs',
      'telemetry.sdk.name': 'opentelemetry',
      'telemetry.sdk.version': '1.25.0'
    }
  },
  traceId: 'a3280f7e6afab1e5b7f4ecfc12ec059f',
  parentId: undefined,
  traceState: undefined,
  name: 'home',
  id: '270509763b408343',
  kind: 0,
  timestamp: 1718975153696000,
  duration: 252.375,
  attributes: { launch: true, state.end: 'active' },
  status: { code: 0 },
  events: [],
  links: []
}
```

> If you dig into the attributes, `launch` refers to the moment the app is launched. It will be `true` only the first time the app mounts. Changing the status between background/foreground won't modify this attribute. For this case the `state.end` is used, and will and it can contain two possible values: `active` and `background`.

Reviewer: ```suggestion
If you dig into the attributes, `launch` refers to the moment the app is launched. It will be `true` only the first time the app mounts. Changing the status between background/foreground won't modify this attribute. For this case the `state.end` is used, and it can contain two possible values: `active` and `background`.
```

Author: fixed

### NOTE

`useProvider` hook in this example returns an instance of a configured provided.

> It doesn't matter which provider you choose; you just need to pass down one (if needed) with all your configurations. To create that provider, you may want to refer to the official [OpenTelemetry JS documentation](https://github.com/open-telemetry/opentelemetry-js). You can also review our suggested implementation (`experimental/testUtils/hooks/useProvider.ts`), but keep in mind that this is the simplest provider with minimal configurations.

Reviewer: ```suggestion
It doesn't matter which provider you choose; you just need to pass down one (if needed) with all your configurations. To create that provider, you may want to refer to the official [OpenTelemetry JS documentation](https://github.com/open-telemetry/opentelemetry-js). You can also review our suggested implementation (`./test/hooks/useProvider.ts`), but keep in mind that this is the simplest provider with minimal configurations.
```

Reviewer: @facostaembrace  I think this one got missed
